File: python/back-end/banco.py
# Conexao com o banco de dados SQLite.
import contextlib
import os
import sqlite3

# Pasta raiz do projeto (um nivel acima de "back-end").
RAIZ = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
CAMINHO_BANCO = os.path.join(RAIZ, "db", "banco.db")
CAMINHO_SCHEMA = os.path.join(RAIZ, "db", "schema.sql")


# Abre uma nova conexao com o banco.
def conectar():
    conexao = sqlite3.connect(CAMINHO_BANCO)
    # O SQLite so respeita as chaves estrangeiras se ligarmos isto em cada conexao.
    conexao.execute("PRAGMA foreign_keys = ON")
    # Espera ate 5s por um lock em vez de falhar com "database is locked".
    conexao.execute("PRAGMA busy_timeout = 5000")
    return conexao


# Verifica se uma conta pertence a um usuario.
def conta_pertence(conta, usuario_id):
    with contextlib.closing(conectar()) as conexao:
        cursor = conexao.cursor()
        cursor.execute(
            "SELECT 1 FROM contas WHERE id = ? AND usuario_id = ?",
            (conta, usuario_id),
        )
        return cursor.fetchone() is not None


# Cria o banco a partir do schema, caso ainda nao exista.
def preparar_banco():
    if not os.path.exists(CAMINHO_BANCO):
        with open(CAMINHO_SCHEMA, encoding="utf-8") as arquivo:
            with contextlib.closing(conectar()) as conexao:
                conexao.executescript(arquivo.read())
                conexao.commit()

File: csharp/back-end/Banco.cs
// Conexao com o banco de dados SQLite.
using Microsoft.Data.Sqlite;

namespace BancoCtf;

public static class Banco
{
    // Caminho do arquivo do banco e do schema (definidos no Program).
    public static string CaminhoBanco = "db/banco.db";
    public static string CaminhoSchema = "db/schema.sql";

    // Abre uma nova conexao com o banco.
    public static SqliteConnection Conectar()
    {
        var conexao = new SqliteConnection($"Data Source={CaminhoBanco}");
        conexao.Open();
        // O SQLite so respeita as chaves estrangeiras se ligarmos isto em cada conexao.
        using var pragma = conexao.CreateCommand();
        pragma.CommandText = "PRAGMA foreign_keys = ON";
        pragma.ExecuteNonQuery();
        return conexao;
    }

    // Cria o banco a partir do schema, caso ainda nao exista.
    public static void CriarSeNaoExiste()
    {
        if (File.Exists(CaminhoBanco))
            return;

        var script = File.ReadAllText(CaminhoSchema);
        using var conexao = Conectar();
        using var cmd = conexao.CreateCommand();
        cmd.CommandText = script;
        cmd.ExecuteNonQuery();
    }

    // Verifica se uma conta pertence a um usuario.
    public static bool ContaPertence(string conta, long usuarioId)
    {
        using var conexao = Conectar();
        using var cmd = conexao.CreateCommand();
        cmd.CommandText = "SELECT 1 FROM contas WHERE id = @conta AND usuario_id = @uid";
        cmd.Parameters.AddWithValue("@conta", conta);
        cmd.Parameters.AddWithValue("@uid", usuarioId);
        return cmd.ExecuteScalar() != null;
    }
}
